bin/oneoff/levelbuilder_source_urls.rb

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env ruby
+
+require_relative '../../dashboard/config/environment'
+require 'cgi'
+require 'uri'
+
+# Toggle writes
+DRY_RUN = true
+
+def fix_start_animations
+  puts DRY_RUN ? "THIS IS A DRY RUN" : "THIS IS A FULL RUN"
+
+  total_levels       = 0
+  modified_levels    = 0
+  urls_updated_total = 0
+  bad_json_levels    = 0
+  # Map: levelbuilder URL (string) -> Set of level names that reference it
+  @levelbuilder_url_to_levels = Hash.new {|h, k| h[k] = Set.new}
+
+  puts "Processing Gamelab levels..."
+  Gamelab.all.each do |level|
+    next unless level.start_animations
+
+    total_levels += 1
+    begin
+      animations = JSON.parse(level.start_animations)
+      ordered_keys = animations['orderedKeys'] || []
+      props_by_key = animations['propsByKey']  || {}
+
+      modified_this_level = false
+
+      ordered_keys.each do |key|
+        props = props_by_key[key]
+        next unless props.is_a?(Hash) && props['sourceUrl'].is_a?(String)
+
+        original_url = props['sourceUrl']
+        new_url = normalize_source_url(original_url, level.name)
+
+        # Only write back if we actually changed the URL string
+        if new_url && new_url != original_url
+          props['sourceUrl'] = new_url
+          urls_updated_total += 1
+          modified_this_level = true
+          # puts "✔ Rewriting on #{level.name}\n  #{new_url}"
+        end
+      end
+
+      if modified_this_level
+        animations['propsByKey'] = props_by_key
+        level.start_animations = JSON.pretty_generate(animations)
+        level.save!(touch: false) unless DRY_RUN
+        modified_levels += 1
+      end
+    rescue
+      puts "Failed to process #{level.name} due to bad JSON"
+      bad_json_levels += 1
+    end
+  end
+
+  if @levelbuilder_url_to_levels.any?
+    puts "\n--- LEVELBUILDER URLS REQUIRING MANUAL REVIEW ---"
+    @levelbuilder_url_to_levels.each do |url, level_names|
+      puts url
+      level_names.sort.each {|name| puts name}
+      puts
+    end
+  end
+
+  puts "\n--- SUMMARY ---"
+  puts "Levels processed:  #{total_levels}"
+  puts "Levels modified:   #{modified_levels}"
+  puts "URLs updated:      #{urls_updated_total}"
+  puts "Bad JSON levels:   #{bad_json_levels}"
+  puts "Levelbuilder URLs: #{@levelbuilder_url_to_levels.size}"
+  puts DRY_RUN ? "NO CHANGES WERE SAVED" : "CHANGES WERE SAVED"
+end
+
+# Returns a possibly-updated URL string, or nil if no change should be made.
+def normalize_source_url(url, level_name)
+  # 1) Recursively unwrap proxies from any domain: .../media?u=<encoded_url>
+  unwrapped = unwrap_proxied_url_recursive(url)
+  uri = safe_uri_parse(unwrapped)
+  return nil unless uri
+
+  unwrapped_changed = (unwrapped != url)
+
+  # 2) If it's in the animationlibrary, make it relative
+  if uri.path&.start_with?('/api/v1/animation-library')
+    return uri.path
+  end
+
+  # 3) Warn if the animation only exists on levelbuilder-studio.code.org
+  if uri.host == 'levelbuilder-studio.code.org'
```

Comment:
for the uris that only exist on levelbuilder, is there a risk of those animations breaking until we manually fix them?

Reply:
Yes! If levelbuilder goes down, they are not available. This has happened the last two Mondays.

Reply:
Oh what I meant was are we changing the urls to not include `levelbuilder` anymore or are those staying the same?